src/d3d9/d3d9_spec_constants.h

```diff
@@ -1,31 +1,119 @@
-    PixelShaderBools  = 8,  // 16 bools               | Bits: 16
-    Fetch4            = 9,  // 1 bit for 16 samplers  | Bits: 16
-                            // ^ not supported for vertex shaders
-
-    SamplerDepthMode  = 10, // 1 bit for 20 samplers  | Bits: 20
-                            // ^ vs + ps
-    SamplerNull       = 11, // 1 bit for 20 samplers  | Bits: 20
-                            // ^ vs + ps
+    SpecSamplerType,        // 2 bits for 16 PS samplers      | Bits: 32
+
+    SpecSamplerDepthMode,   // 1 bit for 20 VS + PS samplers  | Bits: 20
+    SpecAlphaCompareOp,     // Range: 0 -> 7                  | Bits: 3
+    SpecPointMode,          // Range: 0 -> 3                  | Bits: 2
+    SpecVertexFogMode,      // Range: 0 -> 3                  | Bits: 2
+    SpecPixelFogMode,       // Range: 0 -> 3                  | Bits: 2
+    SpecFogEnabled,         // Range: 0 -> 1                  | Bits: 1
+
+    SpecSamplerNull,        // 1 bit for 20 samplers          | Bits: 20
+    SpecProjectionType,     // 1 bit for 6 PS 1.x samplers    | Bits: 6
+
+    SpecVertexShaderBools,  // 16 bools                       | Bits: 16
+    SpecPixelShaderBools,   // 16 bools                       | Bits: 16
+
+    SpecFetch4,             // 1 bit for 16 PS samplers       | Bits: 16
+
+    SpecConstantCount,
+  };
+
+  struct BitfieldPosition {
+    constexpr uint32_t mask() const {
+      return uint32_t((1ull << sizeInBits) - 1) << bitOffset;
+    }
+
+    uint32_t dwordOffset;
+    uint32_t bitOffset;
+    uint32_t sizeInBits;
+  };
+
+  struct D3D9SpecializationInfo {
+    static constexpr uint32_t MaxSpecDwords = 5;
+
+    static constexpr std::array<BitfieldPosition, SpecConstantCount> Layout{{
+      { 0, 0, 32 },  // SamplerType
+      
+      { 1, 0,  20 }, // SamplerDepthMode
+      { 1, 20, 3 },  // AlphaCompareOp
+      { 1, 23, 2 },  // PointMode
+      { 1, 25, 2 },  // VertexFogMode
+      { 1, 27, 2 },  // PixelFogMode
+      { 1, 29, 1 },  // FogEnabled
+
+      { 2, 0,  20 }, // SamplerNull
+      { 2, 20, 6 },  // ProjectionType
+
+      { 3, 0,  16 }, // VertexShaderBools
+      { 3, 16, 16 }, // PixelShaderBools
+
+      { 4, 0,  16 }, // Fetch4
+    }};
+
+    template <D3D9SpecConstantId Id, typename T>
+    bool set(const T& value) {
+      const uint32_t x = uint32_t(value);
+      if (get<Id>() == x)
+        return false;
+
+      constexpr auto& layout = Layout[Id];
+
+      data[layout.dwordOffset] &= ~layout.mask();
+      data[layout.dwordOffset] |= (x << layout.bitOffset) & layout.mask();
+
+      return true;
+    }
+
+    template <D3D9SpecConstantId Id>
+    uint32_t get() const {
+      constexpr auto& layout = Layout[Id];
+
+      return (data[layout.dwordOffset] & layout.mask()) >> layout.bitOffset;
+    }
+
+    std::array<uint32_t, MaxSpecDwords> data = {};
+  };
+
+  class D3D9ShaderSpecConstantManager {
+  public:
+    uint32_t get(SpirvModule &module, D3D9SpecConstantId id) {
```

Comment:
add an offset and range arguments to avoid the double extract for some of the bit fields?

Reply:
Strongly agree.

cc636ed953f50785d42ce52b4fbbb003e91e5f36 should implement this, and also fixes a bug with depth and null samplers being broken in vertex shaders.